Data/Primitive/Array/Class.hs

```diff
@@ -0,0 +1,487 @@
+    readArr = readPrimArray
+    {-# INLINE readArr #-}
+    writeArr = writePrimArray
+    {-# INLINE writeArr #-}
+    setArr = setPrimArray
+    {-# INLINE setArr #-}
+    indexArr = indexPrimArray
+    {-# INLINE indexArr #-}
+    indexArrM arr i = return (indexPrimArray arr i)
+    {-# INLINE indexArrM #-}
+    freezeArr marr s l = do
+        marr' <- newPrimArray l
+        copyMutablePrimArray marr' 0 marr s l
+        unsafeFreezePrimArray marr'
+    {-# INLINE freezeArr #-}
+    thawArr arr s l = do
+        marr' <- newPrimArray l
+        copyPrimArray marr' 0 arr s l
+        return marr'
+    {-# INLINE thawArr #-}
+    unsafeFreezeArr = unsafeFreezePrimArray
+    {-# INLINE unsafeFreezeArr #-}
+    unsafeThawArr = unsafeThawPrimArray
+    {-# INLINE unsafeThawArr #-}
+
+    copyArr = copyPrimArray
+    {-# INLINE copyArr #-}
+    copyMutableArr = copyMutablePrimArray
+    {-# INLINE copyMutableArr #-}
+
+    moveArr = movePrimArray
+    {-# INLINE moveArr #-}
+
+    cloneArr arr s l = runST (do
+            marr <- newPrimArray l
+            copyPrimArray marr 0 arr s l
+            unsafeFreezePrimArray marr
+        )
+    {-# INLINE cloneArr #-}
+    cloneMutableArr marr s l = do
+        marr' <- newPrimArray l
+        copyMutablePrimArray marr' 0 marr s l
+        return marr'
+    {-# INLINE cloneMutableArr #-}
+
+    resizeMutableArr = resizeMutablePrimArray
+    {-# INLINE resizeMutableArr #-}
+    shrinkMutableArr = shrinkMutablePrimArray
+    {-# INLINE shrinkMutableArr #-}
+
+    sameMutableArr = sameMutablePrimArray
+    {-# INLINE sameMutableArr #-}
+    sizeofArr = sizeofPrimArray
+    {-# INLINE sizeofArr #-}
+    sizeofMutableArr = sizeofMutablePrimArray
+    {-# INLINE sizeofMutableArr #-}
+
+    sameArr = samePrimArray
+    {-# INLINE sameArr #-}
+
+instance PrimUnlifted a => Arr MutableUnliftedArray UnliftedArray a where
+    newArr = unsafeNewUnliftedArray
+    {-# INLINE newArr #-}
+    newArrWith = newUnliftedArray
+    {-# INLINE newArrWith #-}
+    readArr = readUnliftedArray
+    {-# INLINE readArr #-}
+    writeArr = writeUnliftedArray
+    {-# INLINE writeArr #-}
+    setArr marr s l x = go s
+      where
+        !sl = s + l
+        go !i | i >= sl = return ()
+              | otherwise = writeUnliftedArray marr i x >> go (i+1)
+    {-# INLINE setArr #-}
+    indexArr = indexUnliftedArray
+    {-# INLINE indexArr #-}
+    indexArrM = indexUnliftedArrayM
+    {-# INLINE indexArrM #-}
+    freezeArr = freezeUnliftedArray
+    {-# INLINE freezeArr #-}
+    thawArr = thawUnliftedArray
+    {-# INLINE thawArr #-}
+    unsafeFreezeArr = unsafeFreezeUnliftedArray
+    {-# INLINE unsafeFreezeArr #-}
+    unsafeThawArr (UnliftedArray arr#) = primitive ( \ s0# ->
+            let !(# s1#, marr# #) = unsafeThawArray# (unsafeCoerce# arr#) s0#   -- ArrayArray# and Array# use the same representation
+            in (# s1#, MutableUnliftedArray (unsafeCoerce# marr#) #)            -- so this works
```

Comment:
I'd like to consult GHC experts here, @bgamari , does my comment hold in this case ? I expect `unsafeThawArray#` works on `ArrayArray#` since they have the same closure type, so this operation will correctly rewrite the closure header. Is that the case?